### PR TITLE
Reference implementation of BIP135 (generalized versionbits) - deprecated by #1088

### PR DESCRIPTION
**Pull request by sanch0panza**
This is an initial reference implementation of my 'generalized versionbits' BIP proposal, the draft specification of which you can find at this link:
https://github.com/bitcoin/bips/blob/master/bip-0135.mediawiki

It provides flexibly configurable parameters for each of the 29 BIP9 deployment bits, including the possibility of

- different thresholds / window sizes
- customizable grace periods (based either on minimum number of blocks or minimum time in seconds elapsed since lock-in)
- explicitly agnostic to fork type (not only intended for soft forks)

The configuration of the deployment bits can be read in from a comma-separated values file, and using the `-dumpforks` option, the built-in configuraiton can also be dumped out in CSV format to standard output.

BIP135 aims to be completely backward compatible (if suitably configured) with the BIP9 state machine. I have taken care to ensure this by leaving the original tests as unchanged as possible, and adding tests for the new functionality in separate scripts / programs.

The "unknown versions" alerting logic in main.cpp needed to be replaced, which slightly alters its behaviour. It now alerts based on a flat > 50/100 threshold and window, and does not attempt to divine the activation states of unknown signals. It also reports on first detection of an unknown signal, as well as gradated (25%, 50%, 70%, 90%, 95%) crossing thresholds, and loss of previously detected unknown signals during a 100-block window.
The 'unknown version' check itself - performed against the computed version according to Unlimited code - should be backward compatible. The notification / alerting code has only been lightly tested, as it is basically optional (not seen as part of core BIP135 requirements) and probably should be implemented in a fancier way, to allow more user control over when/what/how it reports.

In some places this reference implementation is surely a little worse technically than it should be, since I'm not an expert C++ programmer. Please feel free to criticize and offer suggestions for improvement. I'll do my best to provide rationale about any aspects of the implementation.

### CODE REVIEW

**Comment by sanch0panza** [comment]
There are some compiler warnings that my code has introduced which I need to get rid of.

**Comment by sanch0panza** [comment]
Taking this out of WIP for now.

**Comment by sanch0panza** [comment]
Rebased on current 'dev'.
Found I had to make the following changes to test 'bip135-genvbvoting-forks.py' successfully when under heavy load:

```
- time.sleep(3)  # need to actually give daemon a little time to change the state
+ time.sleep(5)  # need to actually give daemon a little time to change the state
```

These changes have not been made in the rebase, to see if Travis tests ok for this without them.

**Comment by sanch0panza** [comment]
Modified PR title and body to reflect BIP number assignment.

**Comment by sanch0panza** [comment]
Rebased on `dev`, did a codebase renaming from 'bip-genvbvoting' to 'bip135'.
Local test of `bip135-genvbvoting-forks.py` OK.
Please note that this test has to be executed manually as it is part of extended testsuite.

I noticed I forgot to add a help text for the new 'bip135_forks' section of 'getblockchaininfo' RPC call. Will add this soon.

File: qa/rpc-tests/bipgenvbvoting-forks.py [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+    def get_bipgenvbvoting_status(self, key):
+        info = self.nodes[0].getblockchaininfo()
+        return info['bipgenvb_forks'][key]
+
+    def print_rpc_status(self):
+        for f in self.defined_forks:
+            info = self.nodes[0].getblockchaininfo()
+            print(info['bipgenvb_forks'][f])
+
+    def test_BIP9_CSV(self, bipName, activated_version, invalidate, invalidatePostSignature, bitno):
```
Comment (dagurval) [inline qa/rpc-tests/bipgenvbvoting-forks.py]: Why are all the tests in `bip9-softforks.py` duplicated here?

Comment (dagurval) [inline qa/rpc-tests/bipgenvbvoting-forks.py]: Why are all the tests in `bip9-softforks.py` duplicated here?

File: qa/rpc-tests/bipgenvbvoting-forks.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+
+        bcinfo = self.nodes[0].getblockchaininfo()
+        # check bits 7-15 , they should be in DEFINED
+        for f in gracebits:
+            assert_equal(bcinfo['bipgenvb_forks'][f]['bit'], int(f[9:]))
+            assert_equal(bcinfo['bipgenvb_forks'][f]['status'], 'defined')
+
+        # move to starttime
+        moved_to_started = False
+        while self.last_block_time < self.fork_starttime or self.height % 10:
```
Comment (dagurval) [inline qa/rpc-tests/bipgenvbvoting-forks.py]: Would make it more readable if you gave `10` a name (change to `self.height % window_size`)

File: qa/rpc-tests/p2p-versionbits-warning.py [hunk 3]
```diff
@@ -73,7 +76,10 @@ def setup_network(self):
         self.nodes.append(start_node(0, self.options.tmpdir, self.node_options))
 
         import re
-        self.vb_pattern = re.compile("^Warning.*versionbit")
+        # bip135 begin
+        # with new unknown version algo, warning has changed.
+        self.vb_pattern = re.compile("^Warning: Unknown block versions being mined Its possible unknown rules are in effect")
```
Comment (dagurval) [inline qa/rpc-tests/p2p-versionbits-warning.py]: Does this pattern really match the warning? Punctuation is missing.

File: qa/rpc-tests/p2p-versionbits-warning.py [hunk 4]
```diff
@@ -14,11 +14,14 @@
...
 
 Generate chains with block versions that appear to be signalling unknown
-soft-forks, and test that warning alerts are generated.
+forks, and test that warning alerts are generated.
 '''
 
-VB_PERIOD = 144 # versionbits period length for regtest
-VB_THRESHOLD = 108 # versionbits activation threshold for regtest
+# bip135 begin
+# modified from 108/144 to 50/100 for new unknown versions algo
```
Comment (dagurval) [inline qa/rpc-tests/p2p-versionbits-warning.py]: Why have these parameters been modified?

File: qa/rpc-tests/p2p-versionbits-warning.py [hunk 5]
```diff
@@ -155,13 +165,13 @@ def run_test(self):
             pass
         self.nodes[0] = start_node(0, self.options.tmpdir, ["-debug", "-logtimemicros=1", "-alertnotify=echo %s >> \"" + self.alert_filename + "\""])
 
-        # Connecting one block should be enough to generate an error.
+        # Since there are no unknown versionbits exceeding threshold in last period,
+        # no error will be generated.
         self.nodes[0].generate(1)
```
Comment (dagurval) [inline qa/rpc-tests/p2p-versionbits-warning.py:175]: A test for error condition was removed here. Is it added somewhere else?

File: src/bitcoind.cpp [hunk 6]
```diff
@@ -97,6 +99,19 @@ bool AppInit(int argc, char* argv[])
         return true;
     }
 
+    // bip135 begin
+    // dump default deployment info and exit, if requested
+    if (mapArgs.count("-dumpforks")) {
```
Comment (dagurval) [inline src/bitcoind.cpp]: Nit: You can use GetBoolArg to avoid accessing mapArgs directly.

File: qa/rpc-tests/bip135-genvbvoting-forks.py [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+        for f in gracebits:
+            assert_equal(bcinfo['bip135_forks'][f]['bit'], int(f[10:]))
+            assert_equal(bcinfo['bip135_forks'][f]['status'], 'defined')
+
+        # move to starttime
+        moved_to_started = False
+        while self.last_block_time < self.fork_starttime or self.height % 10:
+            test_blocks = self.generate_blocks(1, 0x20000000)
+            yield TestInstance(test_blocks, sync_every_block=False)
+            time.sleep(3)  # need to actually give daemon a little time to change the state
```
Comment (dagurval) [inline qa/rpc-tests/bip135-genvbvoting-forks.py]: BIP9 tests do not use `sleep`-hack. Can BIP135 also do without?

File: src/bitcoind.cpp [hunk 8]
```diff
@@ -97,6 +99,19 @@ bool AppInit(int argc, char* argv[])
...
+        std::string strVersion = "# " + strprintf(_("%s Daemon"), _(PACKAGE_NAME)) + " " + _("version") + " " + FormatFullVersion();
+        fprintf(stdout, "%s\n%s", strVersion.c_str(), FORKS_CSV_FILE_HEADER);
+        fprintf(stdout, "%s", NetworkDeploymentInfoCSV(CBaseChainParams::MAIN).c_str());
+        fprintf(stdout, "%s", NetworkDeploymentInfoCSV(CBaseChainParams::UNL).c_str());
+        fprintf(stdout, "%s", NetworkDeploymentInfoCSV(CBaseChainParams::TESTNET).c_str());
+        fprintf(stdout, "%s", NetworkDeploymentInfoCSV(CBaseChainParams::REGTEST).c_str());
+        return true;
```
Comment (dagurval) [inline src/bitcoind.cpp]: Nit: Why not use C++, std::stringstream and std::cout?

File: src/chainparams.cpp [hunk 9]
```diff
@@ -352,14 +358,18 @@ class CRegTestParams : public CChainParams {
...
-        consensus.nRuleChangeActivationThreshold = 108; // 75% for testchains
-        consensus.nMinerConfirmationWindow = 144; // Faster than normal for regtest (144 instead of 2016)
-        consensus.vDeployments[Consensus::DEPLOYMENT_TESTDUMMY].bit = 28;
-        consensus.vDeployments[Consensus::DEPLOYMENT_TESTDUMMY].nStartTime = 0;
-        consensus.vDeployments[Consensus::DEPLOYMENT_TESTDUMMY].nTimeout = 999999999999ULL;
+
         consensus.vDeployments[Consensus::DEPLOYMENT_CSV].bit = 0;
         consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nStartTime = 0;
-        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nTimeout = 999999999999ULL;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nTimeout = 999999999999LL;
```
Comment (dagurval) [inline src/chainparams.cpp:413]: Nit: Why the change from unsigned to signed?

File: src/chainparams.cpp [hunk 10]
```diff
@@ -424,3 +434,69 @@ void SelectParams(const std::string& network)
...
+    const struct BIP9DeploymentInfo& vbinfo = VersionBitsDeploymentInfo[bit];
+
+    if (strlen(vbinfo.name) == 0)
+        return false;
+
+    if (vdeployments[bit].windowsize == 0 || vdeployments[bit].threshold == 0)
+    {
+        return false;
+    }
+    return true;
```
Comment (dagurval) [inline src/chainparams.cpp]: Nit: These 6 lines can be shortened to  `return vdeployments[bit].windowsize != 0 && vdeployments[bit].threshold != 0`

File: src/chainparams.h [hunk 11]
```diff
@@ -113,4 +116,24 @@ CChainParams& Params(const std::string& chain);
...
+CChainParams &ModifiableParams();
+
+/**
+ * Returns true if a deployment is considered active on a particular network
+ */
+
+bool isConfiguredDeployment(const Consensus::Params& consensusParams, const int bit);
+
+/**
+ * Dump the fork deployment parameters for the given BIP70 chain name.
```
Comment (dagurval) [inline src/chainparams.h]: BIP70?

File: src/consensus/params.h [hunk 12]
```diff
@@ -31,6 +67,16 @@ struct BIP9Deployment {
     int64_t nStartTime;
     /** Timeout/expiry MedianTime for the deployment attempt. */
     int64_t nTimeout;
+    // bip135 begin added parameters
```
Comment (dagurval) [inline src/consensus/params.h:72]: Suggestion: Since this struct (`BIP9Deployment`) now serves both BIP9 and BIP135, can we rename it to something more generic? `ForkDeployment`?

**Comment by sanch0panza** [comment]
Please hold off on review while I rebase again and backport some minor improvements from the soon-to-be-released Bitcoin Core PR.

File: src/chainparams.cpp [hunk 13]
```diff
@@ -424,3 +434,69 @@ void SelectParams(const std::string& network)
...
+}
+
+/**
+ * Return a string representing CSV-formatted deployments for the network.
+ * Only configured deployments satisfying isConfiguredDeployment() are included.
+ */
+const std::string NetworkDeploymentInfoCSV(const std::string& network)
```
Comment (dagurval) [inline src/chainparams.cpp]: What's the use-case for this feature? Debugging only?

File: src/versionbits.cpp [hunk 14]
```diff
@@ -1,55 +1,191 @@
...
             cache[pindexPrev] = THRESHOLD_DEFINED;
             break;
         }
+
+        // push the pindex for later forward walking
         vToCompute.push_back(pindexPrev);
+        // go back one more period
         pindexPrev = pindexPrev->GetAncestor(pindexPrev->nHeight - nPeriod);
+
+        if (cache.count(pindexPrev) > 0 && cache[pindexPrev] == THRESHOLD_DEFINED) {
```
Comment (dagurval) [inline src/versionbits.cpp]: Nit: This is the same check as above (line 159). This could be a function to avoid repeating code. You could then call `while (!backAtDefined(cache))` and remove this additional if sentence.

File: src/versionbits.cpp [hunk 15]
```diff
@@ -1,55 +1,191 @@
...
     if (pindexPrev != NULL) {
+        assert(nPeriod);
         pindexPrev = pindexPrev->GetAncestor(pindexPrev->nHeight - ((pindexPrev->nHeight + 1) % nPeriod));
     }
 
-    // Walk backwards in steps of nPeriod to find a pindexPrev whose information is known
+    // Walk backwards in steps of nPeriod to find a pindexPrev which was DEFINED
     std::vector<const CBlockIndex*> vToCompute;
-    while (cache.count(pindexPrev) == 0) {
+    bool backAtDefined = (cache.count(pindexPrev) && cache[pindexPrev] == THRESHOLD_DEFINED);
```
Comment (dagurval) [inline src/versionbits.cpp]: Why did you add the explicit check for state == `THRESHOLD_DEFINED`?

File: src/versionbits.h [hunk 16]
```diff
@@ -16,39 +17,47 @@ static const int32_t VERSIONBITS_TOP_BITS = 0x20000000UL;
...
+    THRESHOLD_LOCKED_IN = 2,
+    THRESHOLD_ACTIVE = 3,
+    THRESHOLD_FAILED = 4,
 };
 
-// A map that gives the state for blocks whose height is a multiple of Period().
-// The map is indexed by the block's parent, however, so all keys in the map
-// will either be NULL or a block with (height + 1) % Period() == 0.
+// A map that gives the state for blocks.
+// The map is indexed by the block's parent.
```
Comment (dagurval) [inline src/versionbits.h]: As far as I can tell this is still caches only blocks whose height is a multiple of Period(), so why the comment change? Or am I missing something?

File: src/main.cpp [hunk 17]
```diff
@@ -2772,44 +2773,111 @@ void static UpdateTip(CBlockIndex *pindexNew) {
...
-        for (int bit = 0; bit < VERSIONBITS_NUM_BITS; bit++) {
-            WarningBitsConditionChecker checker(bit);
-            ThresholdState state = checker.GetStateFor(pindex, chainParams.GetConsensus(), warningcache[bit]);
-            if (state == THRESHOLD_ACTIVE || state == THRESHOLD_LOCKED_IN) {
-                if (state == THRESHOLD_ACTIVE) {
-                    strMiscWarning = strprintf(_("Warning: unknown new rules activated (versionbit %i)"), bit);
-                    if (!fWarned) {
-                        AlertNotify(strMiscWarning);
-                        fWarned = true;
+        // bip135 begin
```
Comment (dagurval) [inline src/main.cpp]: Why do you need to replace the warning code? Why not keep `WarningBitsConditionerChecker`. Also, this big chunk of code would be nice to have in its own function.

**Review by dagurval — COMMENTED** [review]
Why did you add a csv parser? Is it just for the rpc-tests? Did you consider adding the fork data as a configuration argument instead? The software supports multiple arguments with same name, for example:

`bip135=regtest,0,csv,0,999999999999,144,108,108,0,true
bip135=regtest,1,segwit,0,999999999999,144,108,108,0,true
`

That's a lot of work you've done here. Glad to see all the test coverage.

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 4 above)
Comment (sanch0panza) [reply qa/rpc-tests/p2p-versionbits-warning.py]: Because the "unknown versions" checker can no longer rely on 108/144 being the activation threshold of a deployment (the numbers here apply to regtestnet, but the argument applies generally). It is why I had to replace the unknown versions checker code entirely with something more general, and I picked 51/100 as the new threshold at which to warn the user, as that seemed the minimal majority.

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 3 above)
Comment (sanch0panza) [reply qa/rpc-tests/p2p-versionbits-warning.py]: Thanks, this does seem to be an error.

P.S. I will fix this, but also merge corrections I made to this test while porting the whole thing to Core. There were other changes which made the test clearer, such as that the actual error threshold is 51/100 (in this test, 1 is added to some thresholds, which is unnecessarily complicating things).

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 5 above)
Comment (sanch0panza) [reply qa/rpc-tests/p2p-versionbits-warning.py:175]: The removed final check was for an error message that an unknown bit had been activated. This determination is no longer possible when unknown bits have unknown activation thresholds, hence the final test was obsolete.

File: src/bitcoind.cpp (hunk 6 above)
Comment (sanch0panza) [reply src/bitcoind.cpp]: Thanks, will fix

File: src/bitcoind.cpp (hunk 8 above)
Comment (sanch0panza) [reply src/bitcoind.cpp]: Indeed, I should, will fix thanks.

File: src/chainparams.cpp (hunk 9 above)
Comment (sanch0panza) [reply src/chainparams.cpp:413]: nStartTime and nTimeout are signed (int64_t), I thought it better to change the ULL to reflect that.

File: src/chainparams.cpp (hunk 10 above)
Comment (sanch0panza) [reply src/chainparams.cpp]: Thanks, will shorten.

File: src/chainparams.cpp (hunk 13 above)
Comment (sanch0panza) [reply src/chainparams.cpp]: No, not purely debugging - the intended use cases:
- inspect the built-in configuration using the `-dumpforks` command line option (based on a review comment from BIP review that requested: "be self-documenting")
- redirect that to a file which can then be modified to easily create a custom configuration which can be loaded with the `-forks=csvfile` option

File: src/chainparams.h (hunk 11 above)
Comment (sanch0panza) [reply src/chainparams.h]: I cannot yet fully explain the origins of this apparently incorrect BIP reference. It is likely I copied this snippet from another comment either in chainparamsbase.h or chainsparams.h. The earliest introduction I found was with f5ae6c98260fdd3c0ca203ce67c6467d9cdaea95 . How it relates to the actual BIP70 I'm not able to say, but the reference pervades further (rpc/mining.cpp , rpc/blockchain.cpp, qt/networkstyle.h).

File: src/consensus/params.h (hunk 12 above)
Comment (sanch0panza) [reply src/consensus/params.h:72]: That is a sensible renaming which I'd be happy to make.
I will suggest this in the Pull Request to Core too. It would be nice to keep the common name for the data structure.

File: src/main.cpp (hunk 17 above)
Comment (sanch0panza) [reply src/main.cpp]: WarningBitsConditionerChecker was used to check for _activation_ of unknown bits.

With BIP135, it is no longer possible to do that - the best we can do is report on more general events such as those tracked in UnknownForkData . So I had to generalize it, and it seemed useful to add in some thresholds such as 50%, 70%, 95% although they are pretty arbitrary, and the implementation could be reduced somewhat by removing some of them.

I agree that this new checker code should be moved into its own function - will look at improving that.

File: src/versionbits.cpp (hunk 14 above)
Comment (sanch0panza) [reply src/versionbits.cpp]: Thanks, will look at fixing in the manner suggested

File: src/versionbits.cpp (hunk 15 above)
Comment (sanch0panza) [reply src/versionbits.cpp]: It had to do with the need to reliably determine nActualLockinBlock / nActualLockinTime during the ensuing forward walk.

File: src/versionbits.h (hunk 16 above)
Comment (sanch0panza) [reply src/versionbits.h]: You are correct. I very likely modified the comment anticipating that I might lose this property, but it did not turn out that way and I forgot to restore it. Will fix!

**Comment by sanch0panza** [comment]
Thank you @dagurval for the review - I will be making the necessary changes.

>Why did you add a csv parser? Is it just for the rpc-tests?

I wanted a format that could easily be used as a standalone config file. This is of course not obligatory, but it means this feature (fork deployments) can be used without touching other user config. This seemed beneficial to me.

One of the reasons for making it accessible via config was to make testing easier, but another reason, as I mention in another comment above somewhere, was that a reviewer of the original BIP requested it to be "self-documenting", so I thought there should be a good way to inspect the built-in settings, and if one does output them, it should be in a round-trippable format that the user can easily modify and the software can read back in.

> Did you consider adding the fork data as a configuration argument instead? The software supports multiple arguments with same name [...]

No, once I decided on using CSV I no longer looked at mixing with the normal conf file. I wanted to avoid having such config in various places.

I guess you might not have seen https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/458#issuecomment-302928601 before you started your review - sorry about that. I am still planning to make minor changes based on issues I ran into while porting this to Core (PR# https://github.com/bitcoin/bitcoin/pull/10437). Fortunately these do not intersect much with what your review found.

They were things like adding help for the new RPC output, revising the p2p-versionbits-warning.py test to make it clearer what's going on, and also splitting and renaming the bip135-genvbvoting-forks.py into two separate tests for the thresholds and grace periods.

The latter split was mostly out of frustration that I was unable to restart the nodes inbetween the test steps - the Core test framework no longer allowed this in the same way and although I tried, I was unable to figure out how to get it to cooperate.
However, I think having several smaller tests is actually a better way to go.

So there will be another set of changes incoming. I'll mark the ones from your review by their commit message.

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 3 above)
Comment (sanch0panza) [reply qa/rpc-tests/p2p-versionbits-warning.py]: Turned out the actual warning is run through SanitizeString() as part of the alertnotify command processing, that's why the pattern in the test needs to match the sanitized one.

File: qa/rpc-tests/bip135-genvbvoting-forks.py (hunk 7 above)
Comment (sanch0panza) [reply qa/rpc-tests/bip135-genvbvoting-forks.py]: After splitting up the test into two separate ones (as I wanted to anyway), I removed the sleeps and indeed it worked without them. Will remove.

File: src/chainparams.h (hunk 11 above)
Comment (sanch0panza) [reply src/chainparams.h]: Looking at BIP70, it seems it is referring to the 'network' attribute definition which is there given as 

> either "main" for payments on the production Bitcoin network, or "test" for payments on test network. If a client receives a PaymentRequest for a network it does not support it must reject the request.

So I guess BIP70 might have been the first to formally introduce this 'network' concept with the values 'main' and 'test', even if earlier code already used it.
I think it's not a good reference to use, esp. since different clients can custom test networks which are not mentioned by the BIP (and even regtestnet was apparently not around).

So I'll drop the BIP70 reference in newly introduced code.

**Comment by sanch0panza** [comment]
Some more minor refactorings based on review to follow.

Do I need to rebase before Travis will build this branch?

**Comment by sanch0panza** [comment]
@dagurval : I'm not really happy with the factoring out of backAtDefined as a public method, but I could not persuade the compiler while trying to make it a protected method.

If you can find a way to protect it in a way that compiles, I would be most grateful for a patch suggestion.

**Comment by sanch0panza** [comment]
Rebased.

File: src/consensus/params.h (hunk 12 above)
Comment (sanch0panza) [reply src/consensus/params.h:72]: Done - https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/458/commits/08e29dab0eeac942b994465506013fb1a6144e1a

**Comment by sanch0panza** [comment]
Rebased.

**Comment by sanch0panza** [comment]
Rebased.

Passed `make check` and regular regression tests, but seeing extensive local failures of the extended test suite. I will need to compare against `dev` branch without the PR to assess the situation.

Regarding Travis, I cannot see the reason why this build job of 52e9dba failed:
https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/365293103

**Comment by sanch0panza** [comment]
I think the job must have failed because of style checks. I'll run the formatter over the indicated files and try again.

**Comment by ftrader** [comment]
Rebase completed, Travis tests passed.

**Comment by sickpig** [comment]
I personally set up my forked repo to apply format checking automatically, just add this snippets to your `.git/config`:

```
[filter "bitcoin-clang-format"]
        clean = "contrib/devtools/clang-format.py format-stdout-if-wanted clang-format-3.8 %f"
        smudge = cat
```

**Comment by sanch0panza** [comment]
Thank you very much for the advice @sickpig , I will make the changes to my repo!

**Comment by sickpig** [comment]
@sanch0panza need to be rebased.

**Comment by sanch0panza** [comment]
@sickpig @gandrewstone currently clarifying whether there is still intention to integrate this PR

https://bitco.in/forum/threads/buip057-passed-add-bip135-support-bip9.2114/#post-63642

I am willing to rebase close to integration time so that it does not go stale

**Comment by gandrewstone** [comment]
@sanch0panza I'm sorry I dropped the ball on merging this so long ago. Something that I need to work on is when a BUIP is passed, to go back thru PRs looking for any code.  Anyway, now that the fork is complete let's make this happen.  If you would do a rebase, that would be great!

**Comment by sanch0panza** [comment]
Thanks @gandrewstone , I appreciate you're all busy with the fork and upgrades that happened. 'll get on rebasing this one, should be pushed later today. CC: @sickpig

**Comment by sanch0panza** [comment]
Hmm - I rebased and pushed with `-f` to my personal branch at

https://github.com/sanch0panza/bitcoin/tree/genvbvoting-bu-dev-clean1

I can see the commits there, but this PR isn't updated as I thought it would be.
Will take a look at this a bit later, right now I can't think of what I did differerently to previous rebases.

**Comment by ptschip** [comment]
@sanch0panza there are still conflicts with three files.  You maybe didn't download the current dev before rebasing?

**Comment by sanch0panza** [comment]
@ptschip @gandrewstone : Before my previous comment, I pulled the latest 'dev', and rebased on that. Those three files had to be modified (it was just file header comments that needed adapting to the latest head with copright date 2018).

The rebase completed, and I pushed with `-f` to my repo. Usually at this point github will update the PR, but this didn't happen this time round.
Now when I pull `dev` to my local equivalent it tells me I'm up to date.
No rebase in progress anymore. When I push to my work branch, it tells me everything is up to date.

**Comment by sanch0panza** [comment]
I tried two things now, which didn't work.

1. follow this recipe to add a temporary file to my branch and force push again: https://makandracards.com/makandra/13043-force-github-pull-requests-to-update-the-diff-against-its-target-branch

That worked in the sense that git gave me no errors, but it didn't cause github to update the PR either.

2. I used github's GUI "Resolve conflicts" to resolve the 3 conflicting files. However, after I've completed all of them, and click on "Commit merge" to commit the resolution, the GUI indicates there is an error "This page is out of date" and that I need to start again. Doing so leads to the same situation again.

What I notice is that on `dev`, chainparams.cpp starts with a BOM. Not sure if that is in any way relevant. It is the only file in `src/` to do so.

**Comment by sickpig** [comment]
@sanch0panza I think you did the right thing, i.e. rebased and pushed -f
Usually this is enough to have the PR updated. 

Try this on your local repo

```sh
 git commit --amend --no-edit && git push -f
```

WRT BOM: no source file should start with a byte order mark, it has to be a left over and we need to remove it

**Comment by sanch0panza** [comment]
@sickpig: I added a commit https://github.com/sanch0panza/bitcoin/commit/437a1de791b90c9f83a486a75f1012802288f7b3 to remove the BOM in my branch and did a `push -f`. I noticed github web page turns to 'Checking for conflicts' momentarily, but it goes back to 'This branch has conflicts that must be resolved' which we can see now.

Then I did your advice to force another update. Same behavior again - github obviously processes something, but it thinks my branch has conflicts.

Since that didn't work, I did another rebase, but it's still the same. My branch in github is up to date, this PR thinks it still has conflicts.

**Comment by sickpig** [comment]
Guess this is due to some merge problems we had during the 1.3.0.x release cycle. What about opening a new PR just to check if the conflicts persist also in the new one?

**Comment by sanch0panza** [comment]
Yes, I'm out of other ideas, so I'll try opening a new PR. So far it's proving difficult.

I can't select BU's repo and 'dev' branch as the base in the github GUI dropdowns, and constructing a compare URL like this by hand also doesn't yield a working comparison:

https://github.com/bitcoinunlimited/bitcoinunlimited/compare/dev...sanch0panza:genvbvoting-bu-dev-clean1

I'm probably going to have to create another repo in my account, forked off BU specifically, to create a PR.

**Comment by sickpig** [comment]
ok, I think this should work

**Comment by sickpig** [comment]
@sanch0panza do you think we could close this one now that we have #1088?

**Comment by sanch0panza** [comment]
@sickpig Yes this one can be closed...

**Comment by sickpig** [comment]
closing in favor of #1088
